### ISSUE TITLE
PeptideIndexer: "unknown AA" error due to selenocysteine

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#2594]
Since #2545, PeptideIndexer may write error messages like this:
> Error while parsing sequence MKKVV...: found an unknown AA without an estimated mass. Please use PEPTIX[123] syntax to indicate an unknown amino acid with a known mass.

While getting this information now is probably a step forward, there are some problems:
- The consequences of the error are unclear, since the tool finishes normally. Are the affected sequences ignored? Is it just a warning?
- The protein sequence that is written may be very long, in which case it isn't easy to see where the problematic character is.
- The sequence that is written may not occur exactly like that in the FASTA file: If "IL_equivalent" is set, "L" will be replaced by "I"; "invalid" letters (e.g. "U") are replaced by "X". (So if I search for "X" in the FASTA file, I may not find anything.)
- In my case, the problematic letter in the original sequence was a "U", which stands for selenocysteine and is actually correct  for the protein in question (E. coli formate dehydrogenase). The fact that OpenMS (or just PeptideIndexer?) doesn't support this points to a more general problem.

**Comment by hendrikweisser** [comment]
Also, since PeptideIndexer explicitly has functionality for dealing with ambiguous amino acids like "X" (e.g. parameter "aaa_max"), their presence should not lead to errors.

**Comment by jpfeuffer** [comment]
A problem might be that we actually use AASequence objects in PeptideIndexer. In the PR we all assumed that during DB handling this would not be the case.
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/ANALYSIS/ID/PeptideIndexing.cpp#L157
I think it is not necessary to use an AASequence object here, is it?

**Comment by hendrikweisser** [comment]
@jpfeuffer: Nice analysis. More generally, we should check whether the functions in EnzymaticDigestion couldn't be made to operate on Strings instead of AASequences. If they only use the sequence information and don't need access to chemical information etc., then I think we should avoid the costly - and potentially tricky - conversion to AASequence.

**Comment by timosachsenberg** [comment]
I agree but would like to hear @cbielow on this. AFAIK he/we wanted to support digestion of fasta files with modifications

**Comment by cbielow** [comment]
Strings in EnzymaticDigestion should work just nicely in the current context.
So I'd say do it.

**Comment by timosachsenberg** [comment]
Maybe just use that StringView based version I wrote a while ago?
    /// Performs the enzymatic digestion of a unmodified protein String. By returning only references into the original string it is much faster. max_length restricts the maximum length of reported peptides (0 = no restriction)
    void digestUnmodifiedString(const StringView sequence, std::vector<StringView>& output, Size min_length = 1, Size max_length = 0) const;

**Comment by timosachsenberg** [comment]
it is already in EnzymaticDigestion and does not require copying strings

**Comment by timosachsenberg** [comment]
Why is this the error happening anyways:
We have:
```
		<NODE name="Selenocysteine">
			<ITEM name="Name" value="Selenocysteine" type="string" />
			<ITEM name="ShortName" value="Sec" type="string" />
			<ITEM name="ThreeLetterCode" value="SEC" type="string" />
			<ITEM name="OneLetterCode" value="U" type="string" />
			<ITEM name="Formula" value="C3H7NO2Se" type="string" />
			<ITEM name="pkc" value="5.73" type="string" />
			<ITEM name="GB_SC" value="0.00" type="string" />
			<ITEM name="GB_BB_L" value="880.99" type="string" />
			<ITEM name="GB_BB_R" value="0.12" type="string" />
			<ITEM name="ResidueSets" value="All,AmbiguousWithoutX,Ambiguous,AllNatural" type="string" />
			<NODE name="Synonyms">
				<ITEM name="Sec" value="Sec" type="string" />
			</NODE>
		</NODE>
```

**Comment by timosachsenberg** [comment]
fixed according to https://github.com/OpenMS/OpenMS/pull/2685
Close? @cbielow

**Comment by cbielow** [comment]
jep. This is fixed, since we do not instanciate AASequence.
The downside is of course that any FASTA file(!) (not idXML!) which contains modifications in the sequence will be parsed 'as is'. This means that peptide hits in the affected regions will be overlooked (because the RegEx engine will not skip over modifications). But that is an extreme case, which should not happen in the first place.
Or should we introduce a critical error in this case? What characters should we look for? Is `[` enough?

**Comment by timosachsenberg** [comment]
I would be fine with an error (or warning) for ( and [

Others (e.g. numbers) are skipped?
quoting wikipedia:
```
Sequences may be protein sequences or nucleic acid sequences, and they can contain gaps or alignment characters. ... Lower-case letters are accepted and are mapped into upper-case; a single hyphen or dash can be used to represent a gap character; and in amino acid sequences, U and * are acceptable letters (see below). Numerical digits are not allowed but are used in some databases to indicate the position in the sequence.
```

**Comment by cbielow** [comment]
fixed in #2685 (and covered even more cases in #3046)
